### PR TITLE
Convert gasnet-ex shim barrier over to the GEX API

### PR DESCRIPTION
**Pull request by ronawho**
Convert barrier over to the GEX API. GASNet-EX barriers are anonymous and use the new collective interface and standard GEX events to check for completion. Note that unlike `gasnet_barrier_try`, `gex_Event_Test` does not poll, so here we need to manually poll if we're barrier'ing early in init before the progress thread has been started or late in deinit after the progress thread has stopped.

Since barriers are now anonymous, it's not possible to receive a barrier mismatch error, which I think was the only interesting error case before. This enables us to remove the IMO confusing error checking that was done before (confusing because the macro looked like we were making another barrier call, even though we were just using the string for error messages.)

Testing:
 - [ ] gex-udp-everything: full suite
 - [ ] gex-ibv-large: `release/examples runtime/configMatters`
 - [ ] barrier perf test 

Part of #14443

### CODE REVIEW

**Comment by ronawho** [comment]
@bonachea @PHHargrove I'm opening this as a draft since it's hanging for ibv testing. The current code is not polling during the barrier wait loop under the assumption that the `RCV_THREAD` would handle any polling that's required. Is that an incorrect assumption?

**Comment by bonachea** [comment]
> @bonachea @PHHargrove I'm opening this as a draft since it's hanging for ibv testing. The current code is not polling during the barrier wait loop under the assumption that the `RCV_THREAD` would handle any polling that's required. Is that an incorrect assumption?

ibv-conduit defaults to `GASNET_BARRIER=RDMADISSEM` which performs barrier coordination using RDMA writes. Those writes do not wake up the `ibv-rcv-thread`, which only reliably services AMs incoming from the InfiniBand network. As such it makes sense you might see hangs using that barrier when only the recv thread is polling. If I'm correct then setting `GASNET_BARRIER=AMDISSEM` will likely hide the problem, although it might still suffer from hangs with multi-PPN due to missing incoming AM messages on the shared-memory channel.

TL:DR: I think you've identified a second deficiency in relying exclusively on the `ibv-rcv-thread` to perform GASNet polling.

File: runtime/src/comm/gasnet/comm-gasnet-ex.c [hunk 1]
```diff
@@ -1084,23 +1066,15 @@ void chpl_comm_broadcast_private(int id, size_t size) {
...
-  while ((retval = gasnet_barrier_try(id, 0)) == GASNET_ERR_NOT_READY) {
+  gex_Event_t bar = gex_Coll_BarrierNB(myteam, GEX_NO_FLAGS);
+  while (gex_Event_Test(bar) != GASNET_OK) {
+    // Need to poll if polling is required and progress thread isn't running
+    // (polling thread is started late in init and stopped early in deinit)
+    if (pollingRequired && !pollingRunning) {
+      gasnet_AMPoll();
+    }
     chpl_task_yield();
   }
```
Comment (bonachea) [inline runtime/src/comm/gasnet/comm-gasnet-ex.c:1077]: I suggest polling unconditionally in this loop:

```suggestion
  gex_Event_t bar = gex_Coll_BarrierNB(myteam, GEX_NO_FLAGS);
  while (gex_Event_Test(bar) != GASNET_OK) {
    // ibv-conduit polling thread alone does not reliably advance barriers,
    // so poll here unconditionally
    gasnet_AMPoll();
    chpl_task_yield();
  }
```

This will probably fix your hang, although it might increase lock contention and hence injection overhead for concurrent threads.

**Review by bonachea — CHANGES_REQUESTED** [review]

File: runtime/src/comm/gasnet/comm-gasnet-ex.c (hunk 1 above)
Comment (ronawho) [reply runtime/src/comm/gasnet/comm-gasnet-ex.c:1077]: Yeah, an unconditional poll resolves the hang. I was hoping to avoid that with the goal of reducing contention, but seems like it's needed. I expect to use our `am_poll_try` wrapper though.

Do you have a sense for what about the RDMA-based barrier requires AM progress calls?

Comment (bonachea) [reply runtime/src/comm/gasnet/comm-gasnet-ex.c:1077]: > Do you have a sense for what about the RDMA-based barrier requires AM progress calls?

It's not that it requires AM progress calls, because it doesn't use AM. `gasnet_AMPoll()` has a slightly misleading name because it also polls the GASNet barrier subsystem (along with a few other GASNet subsystems), and when RDMADISSEM barrier is active that includes checking whether any incoming RMA Puts have arrived to advance the barrier state.

**Comment by ronawho** [comment]
FYI I see a slight performance regression switching to the GEX barrier (this PR plus using an unconditional `am_poll_try()` or `gasnet_AMPoll`). Results for a barrier microbenchmark on a 72-core Icelake system with HDR-200 IB (with similar results on a 128-core Milan system):


```chpl
use Time, CTypes;
config const numTrials = 500_000;


inline proc barrier() {
  extern proc chpl_comm_barrier(msg: c_ptrConst(c_char));
  chpl_comm_barrier('b');
}

var s: stopwatch; s.start();
coforall loc in Locales do on loc do
  for 1..numTrials do
    barrier();
writeln(s.elapsed());
```

```sh
CHPL_GASNET_VERSION=1  chpl bar.chpl --fast --no-cache-remote -o bar-g1
CHPL_GASNET_VERSION=ex chpl bar.chpl --fast --no-cache-remote -o bar-gex

./bar-g1 -nl 16
2.43624

./bar-gex -nl 16
2.62389
```

File: runtime/src/comm/gasnet/comm-gasnet-ex.c (hunk 1 above)
Comment (ronawho) [reply runtime/src/comm/gasnet/comm-gasnet-ex.c:1077]: Understood, thanks. And no current or expected interface to progress specific subsystems (I believe know, but want to verify).

Already captured as "explore which functions poll internally" in https://github.com/chapel-lang/chapel/issues/14443, but I think it'll be important to consider where progress calls are needed as some things are converted to use events (and as we look to make Blocking RMA calls into non-blocking ones)

Comment (bonachea) [reply runtime/src/comm/gasnet/comm-gasnet-ex.c:1077]: > And no current or expected interface to progress specific subsystems (I believe no, but want to verify).

Correct.

We expect there will eventually be a mechanism to progress a specific set of endpoints, but we're not there yet (and currently Chapel is only using one endpoint per process).

**Comment by bonachea** [comment]
> FYI I see a slight performance regression switching to the GEX barrier (this PR plus using an unconditional `am_poll_try()` or `gasnet_AMPoll`). Results for a barrier microbenchmark on a 72-core Icelake system with HDR-200 IB (with similar results on a 128-core Milan system):

This is slightly concerning. How reproducible is this result, could it be due to noise? Is the job mapping to 16 physical nodes with 1 PPN? Did you use core pinning to ensure the processes are not freely floating around between NUMA domains? I'm also curious how `export GASNET_BARRIER=AMDISSEM` changes the results.

The Chapel program you tested doesn't seem to spin up multiple tasks, so IIUC thread contention should not even be an issue in this benchmark: ie. the only threads that should be running are the single application thread and the ibv-conduit rcv thread (which IIUC should sleep entirely through the timed region when using an RDMA barrier). I don't have a good theory to explain the difference you measured, but perhaps @PHHargrove has some ideas?

**Comment by ronawho** [comment]
It's consistently reproducible on two different systems, so I don't think noise. The job is mapping to 16 physical nodes with 1PPN and threads are pinned to cores. This test is just running a single task on a single thread so like you say shouldn't be any contention and should just be this one thread active, other chapel threads blocked on a condition variable, and the rcv thread blocked on whatever it blocks on.

I'll try with the AM barrier and see if I can try with a standalone gasnet test.

**Comment by ronawho** [comment]
`AMDISSEM` results:

| Conf       | G1 Barrier | Gex Barrier |
| ---------- | ---------: | ----------: |
| RDMADISSEM | 2.46483    | 2.66273     |
|   AMDISSEM | 4.9535     | 6.33736     |

**Comment by PHHargrove** [comment]
@ronawho I will try to get a look at what might be going on as soon as I can, but am at SC right now and fly back home tonight.   Time at my gate in the Denver airport might be my first good opportunity.   In the meantime, can you please quantify "a slight performance regression"?

**Comment by ronawho** [comment]
The table in https://github.com/chapel-lang/chapel/pull/23845#issuecomment-1813275358 quantifies the regression (~2.4s -> ~2.6s for g1 to gex barrier with rdma barrier, ~4.9s -> ~6.3s for AM)

And I collected some standalone gasnet results with testbarrier:

```
./ibv-conduit/contrib/gasnetrun_ibv -n 16 -N 16 -c 1 ./ibv-conduit/testbarrier -p 0 500000
=====> testbarrier nprocs=16 config=RELEASE=2023.3.0,SPEC=0.16,CONDUIT=IBV(IBV-2.12/IBV-2.12),THREADMODEL=PAR,SEGMENT=LARGE,PTR=64bit,CACHE_LINE_BYTES=64,noalign,nopshm,nodebug,notrace,nostats,nodebugmalloc,nosrclines,timers_native,membars_native,atomics_native,atomic32_native,atomic64_native,notiopt compiler=CLANG/15.0.4  sys=x86_64-pc-linux-gnu
Running barrier test with 500000 iterations and 0 extra polling threads...
...
Total time:    2.760 sec  Avg Named Barrier latency:    5.520 us
Total time:    2.759 sec  Avg Anon. Barrier latency:    5.518 us
Total time:    2.764 sec  Avg Mixed Barrier latency:    5.528 us
Total time:    2.758 sec  Avg Uname Barrier latency:    5.517 us
Total time:    2.772 sec  Avg GEX Blocking Barrier latency:    5.543 us
Total time:    2.886 sec  Avg GEX NB Barrier latency:    5.773 us
...
```

Where the raws times aren't exactly the same as what I see in chapel, but the GEX NB latency is higher compared to the named G1 barrier.

**Comment by PHHargrove** [comment]
Found some time sooner than expected.

Consistent with your report, @ronawho , below are results from `make testbarrier-par ; ./contrib/gasnetrun_ibv -n 2 ./testbarrier 1000000` on an Intel Skylake system with EDR InfiniBand.  This is a thread-safe (PAR) build but the microbenchbark does not spawn any threads.  There is a clearly visible cost for using the GEX APIs (last two results) versus the G-1 APIs (first four) even for a 2-node run.

```
=====> testbarrier nprocs=2 config=RELEASE=2023.3.3,SPEC=0.16,CONDUIT=IBV(IBV-2.12/IBV-2.12),THREADMODEL=PAR,SEGMENT=FAST,PTR=64bit,CACHE_LINE_BYTES=64,noalign,pshm,nodebug,notrace,nostats,nodebugmalloc,nosrclines,timers_native,membars_native,atomics_native,atomic32_native,atomic64_native,notiopt compiler=GNU/12.1.0 sys=x86_64-pc-linux-gnu
node 0/2 hostname is: skylake09 (supernode=0 pid=20395)
node 1/2 hostname is: skylake06 (supernode=1 pid=41619)
Running barrier test with 1000000 iterations and 0 extra polling threads...

Total time:    1.655 sec  Avg Named Barrier latency:    1.655 us
Total time:    1.654 sec  Avg Anon. Barrier latency:    1.654 us
Total time:    1.661 sec  Avg Mixed Barrier latency:    1.661 us
Total time:    1.654 sec  Avg Uname Barrier latency:    1.654 us
Total time:    1.718 sec  Avg GEX Blocking Barrier latency:    1.718 us
Total time:    1.936 sec  Avg GEX NB Barrier latency:    1.936 us
node 0/2 done.
node 1/2 done.
```

A SEQ (no thread safety) build shows smaller penalties, but the GEX NB Barrier is still the slowest:
```
Total time:    1.644 sec  Avg Named Barrier latency:    1.644 us
Total time:    1.641 sec  Avg Anon. Barrier latency:    1.641 us
Total time:    1.640 sec  Avg Mixed Barrier latency:    1.640 us
Total time:    1.644 sec  Avg Uname Barrier latency:    1.644 us
Total time:    1.658 sec  Avg GEX Blocking Barrier latency:    1.658 us
Total time:    1.870 sec  Avg GEX NB Barrier latency:    1.870 us
```

So, this is not specific to your systems.

Runs with `GASNET_BARRIER=AMDISSEM` and the PAR build are qualitatively similar:
```
Total time:    1.755 sec  Avg Named Barrier latency:    1.755 us
Total time:    1.751 sec  Avg Anon. Barrier latency:    1.751 us
Total time:    1.752 sec  Avg Mixed Barrier latency:    1.752 us
Total time:    1.751 sec  Avg Uname Barrier latency:    1.751 us
Total time:    1.806 sec  Avg GEX Blocking Barrier latency:    1.806 us
Total time:    1.986 sec  Avg GEX NB Barrier latency:    1.986 us
```

I don't have any profiling to support it, but my guess is that thread-safety overheads, in particular those associated with event management, it likely responsible for the difference between G-1 and GEX.  Running a PAR build of `testmisc` reports 0.016 us for cycling an uncontended mutex.  So that alone cannot account for a difference more than 10x larger unless there are more mutex acquisitions in the critical paths than I would imagine.

I _might_ find time later to walk myself through the code to look for clues where this time difference comes from.

**Comment by PHHargrove** [comment]
@ronawho I expect to find time in December to work through the reason(s) of the GEX barrier is so much slower than the G-1 version. Dan and I are now certain it is a consequence of the added generality of allowing it to mix freely with other collectives.  However, it is unclear if we can modify the implementation to provide that generality less expensively.
Until I am able to work on this, please don't expend any more of _your_ effort on this.

**Comment by ronawho** [comment]
Understood. Do you have a preference between merging this code and accepting the regression for now and waiting for your investigation? (I lean towards merging personally)

**Comment by PHHargrove** [comment]
As guessed in the "edit:" above, I request that you wait on us to make reasonable attempts to address the problem rather than moving forward with a merge.  I suspect @bonachea will speak up if he has other thoughts (since he has more background on this pull request than I do).

Regarding "of the types of things the new interface enables", here is the relevant `GASNet-EX.txt` text which basically acknowledges that the G-1 barrier (and the GEX one in old releases) could not be mixed with other collective operations, nor could one have multiple barriers in-flight per team (but both are possible now):
```
// This API is an updated and expanded version of the collectives prototype
// offered in GASNet-1, and previously documented in docs/collective_notes.txt.
// As these APIs are fully specified and implemented, the corresponding
// portions of the GASNet-1 collectives prototype will be removed from
// gasnet_coll.h and replaced with GEX variants.  The GASNet-1 collectives API
// signatures will not be supported in future releases.

// The following semantics apply to all Coll functions:

// For NB variants, return type for all functions in this section is gex_Event_t.
// There are no NBI or Blocking variants at this time.

// All functions in this section are "Collective Calls" as defined in the
// Glossary.

// Multiple collective operations from this section may be active concurrently,
// over multiple teams or over a single team.  There is no longer an exception
// regarding gex_Coll_BarrierNB(), as was the case in an earlier release.
```

**Comment by ronawho** [comment]
Understood on waiting. 

Do you have an example of multiple collectives done by a single team? Could you do consecutive broadcasts without a wait? Have distinct threads participating in distinct barriers?
